File: frontend/src/types/matching.ts
export type MatchStatus = "PENDING" | "RUNNING" | "SUCCESS" | "FAILED" | "CANCELLED"

export type MatchPhase =
  | "VALIDATING_INPUT"
  | "EXTRACTING_REQUIREMENTS"
  | "NORMALIZING_SKILLS"
  | "MATCHING_RULES"
  | "MATCHING_SEMANTIC"
  | "VERIFYING_EVIDENCE"
  | "CALCULATING_SCORE"
  | "SAVING_REPORT"
  | "COMPLETED"

export type SkillMatchStatus = "MATCHED" | "PARTIAL" | "MISSING" | "UNKNOWN"
export type AvailabilityStatus = "PROVIDED" | "UNKNOWN" | "NOT_PROVIDED"
export type RiskSeverity = "INFO" | "LOW" | "MEDIUM" | "HIGH" | "BLOCKING"
export type RiskType = "SKILL_GAP" | "BLOCKING_RISK" | "GENERAL_RISK"
export type VerificationStatus = "CONFIRMED" | "UNVERIFIED" | "CONFLICT"
export type DimensionStatus = "GOOD" | "PARTIAL" | "WEAK" | "UNKNOWN"
export type RecommendationLevel =
  | "STRONGLY_RECOMMENDED"
  | "RECOMMENDED"
  | "CONSIDER"
  | "HIGH_RISK"
  | "NOT_RECOMMENDED"
export type ReportConfidence = "HIGH" | "MEDIUM" | "LOW" | "VERY_LOW"
export type ScoringVersion = "deterministic-v1" | "deterministic-v1.1" | "hybrid-v1"

export interface ScoreWeights {
  hard_skills: number
  evidence_strength: number
  experience_education: number
  language_location_eligibility: number
  user_preferences: number
  other_conditions: number
}

export interface ScoringConfig {
  version: string
  skill_dictionary_version: string
  job_requirement_extractor_version: string
  weights: ScoreWeights
  partial_skill_credit: number
  unknown_criterion_credit: number
  required_skill_weight: number
  preferred_skill_weight: number
  base_rule_version?: string | null
  semantic?: SemanticScoringConfig | null
  hybrid_weights?: HybridWeights | null
  embedding_config?: EmbeddingConfigSnapshot | null
}

export interface SemanticScoringConfig {
  version: string
  top_k: number
  similarity_threshold: number
  similarity_ceiling: number
  empty_score: number
}

export interface HybridWeights {
  deterministic: number
  semantic: number
}

export interface EmbeddingConfigSnapshot {
  provider: string
  model: string
  device: string
  normalize_embeddings: boolean
}

export interface SemanticEvidence {
  rank: number
  relation: string
  similarity: number
  score: number
  resume_text: string
  job_text: string
  resume_section: string
  job_field: string
  page_number: number | null
  paragraph_index: number | null
  resume_metadata: Record<string, string | number | boolean | null>
  job_metadata: Record<string, string | number | boolean | null>
}

export interface SemanticSummary {
  evidence_count: number
  evaluated_resume_chunks: number
  top_similarity: number | null
  relation_counts: Record<string, number>
}

export interface DimensionScore {
  code: string
  label: string
  score: number
  weight: number
  weighted_score: number
  explanation: string
  status: DimensionStatus
  evidence_keys: string[]
  gap_codes: string[]
}

export interface EvidenceTrace {
  conclusion_key: string
  resume_version_id: number
  resume_skill_id: number | null
  resume_evidence: string | null
  resume_section: string | null
  source_type: string | null
  page_number: number | null
  paragraph_index: number | null
  source_location: string | null
  job_field: string
  job_snippet: string
  matching_rule: string
  confidence: number
}

export interface SkillMatch {
  normalized_name: string
  job_raw_name: string
  category: string | null
  requirement_type: string
  status: SkillMatchStatus
  score: number
  matching_rule: string
  explanation: string
  resume_skill_id: number | null
  evidence_count: number
  source_types: string[]
  responsibility_keyword_overlap: boolean
}

export interface RiskItem {
  code: string
  title: string
  severity: RiskSeverity
  explanation: string
  job_requirement: string | null
  resume_evidence: string | null
  remediation: string
  risk_type: RiskType
  verification_status: VerificationStatus
}

export interface RecommendedAction {
  code: string
  title: string
  explanation: string
  priority: RiskSeverity
}

export interface JobSkillRequirement {
  raw_name: string
  normalized_name: string
  category: string | null
  required: boolean
  weight: number
  evidence_text: string
  source_field: string
  normalization_rule: string
  confidence: number
}

export interface JobRequirements {
  job_title: string
  industry: string | null
  salary_min: number | null
  salary_max: number | null
  skills: JobSkillRequirement[]
  minimum_experience_years: number | null
  experience_status: AvailabilityStatus
  graduate_exception: boolean
  education_level: string | null
  education_status: AvailabilityStatus
  languages: string[]
  language_status: AvailabilityStatus
  location: string | null
  location_status: AvailabilityStatus
  employment_type: string | null
  employment_status: AvailabilityStatus
  certificates: string[]
  certificates_required: boolean
  certificate_status: AvailabilityStatus
  work_eligibility: string | null
  work_eligibility_status: AvailabilityStatus
  experience_required_explicit: boolean
  education_required_explicit: boolean
  education_non_substitutable: boolean
  language_required_explicit: boolean
  language_preferred: boolean
  certificate_required_explicit: boolean
  certificate_preferred: boolean
  warnings: string[]
  extractor_version: string
}

export interface EvidenceCoverage {
  total_requirements: number
  verified: number
  partially_supported: number
  missing: number
  unknown: number
  coverage_percent: number
}

export interface MatchInputSnapshot {
  resume: {
    resume_id: number
    version_id: number
    version_number: number
    is_confirmed: boolean
    created_at: string
  }
  job: {
    job_id: number
    title: string
    company: string
    location: string | null
    content_hash: string
    import_method: string | null
    description: string
    requirements: string | null
    responsibilities: string | null
  }
}

export interface MatchReport {
  id: number
  user_id: number
  resume_version_id: number
  job_id: number
  status: MatchStatus
  current_phase: MatchPhase
  rule_score: number | null
  semantic_score: number | null
  hybrid_score: number | null
  final_score: number | null
  scoring_version: ScoringVersion
  scoring_config_snapshot: ScoringConfig
  dimension_scores: DimensionScore[]
  matched_skills: SkillMatch[]
  partial_skills: SkillMatch[]
  missing_skills: SkillMatch[]
  hard_requirement_risks: RiskItem[]
  evidence_items: EvidenceTrace[]
  recommendation: RecommendationLevel | null
  explanation: string | null
  recommended_actions: RecommendedAction[]
  job_requirements: JobRequirements | null
  phase_timings_ms: Record<string, number>
  duration_ms: number | null
  error_code: string | null
  error_message: string | null
  started_at: string | null
  completed_at: string | null
  created_at: string
  updated_at: string
  job_information_completeness?: number | null
  report_confidence?: ReportConfidence | null
  report_confidence_score?: number | null
  completeness_warning?: string | null
  evaluated_requirement_count?: number
  unknown_requirement_count?: number
  not_provided_requirement_count?: number
  recommendation_cap?: RecommendationLevel | null
  policy_version?: string | null
  policy_config_snapshot?: Record<string, unknown> | null
  embedding_model?: string | null
  embedding_config_snapshot?: EmbeddingConfigSnapshot | null
  semantic_config_snapshot?: SemanticScoringConfig | null
  semantic_evidence?: SemanticEvidence[]
  semantic_summary?: SemanticSummary | null
  evidence_coverage: EvidenceCoverage
  input_snapshot?: MatchInputSnapshot | null
}

export interface MatchCreateInput {
  resume_version_id: number
  job_id: number
  scoring_version?: ScoringVersion
}

export interface MatchCreateResult {
  report: MatchReport
  reused: boolean
}

export interface MatchList {
  items: MatchReport[]
  total: number
  offset: number
  limit: number
}

export interface MatchStatusResult {
  id: number
  status: MatchStatus
  current_phase: MatchPhase
  phase_timings_ms: Record<string, number>
  duration_ms: number | null
  error_code: string | null
  error_message: string | null
}

export type MatchRunStatus =
  | "PENDING"
  | "RUNNING"
  | "WAITING_REVIEW"
  | "SUCCEEDED"
  | "FAILED"
  | "CANCELLED"
export type MatchNodeStatus =
  | "PENDING"
  | "RUNNING"
  | "SUCCEEDED"
  | "FAILED"
  | "SKIPPED"
  | "WAITING_REVIEW"
  | "CANCELLED"

export interface MatchRunStep {
  node_name: string
  status: MatchNodeStatus
  retry_count: number
  started_at: string | null
  finished_at: string | null
  duration_ms: number | null
  summary: Record<string, unknown>
  error_message: string | null
}

export interface MatchRun {
  run_id: number
  user_id: number
  resume_version_id: number
  job_id: number
  scoring_version: ScoringVersion
  status: MatchRunStatus
  current_node: string | null
  node_status: MatchNodeStatus
  completed_nodes: string[]
  rule_score: number | null
  semantic_score: number | null
  hybrid_score: number | null
  blocking_risks: Array<Record<string, unknown>>
  report_id: number | null
  error_code: string | null
  error_message: string | null
  waiting_for_user: boolean
  retry_count: number
  started_at: string | null
  finished_at: string | null
  duration_ms: number | null
  steps: MatchRunStep[]
}

export interface MatchRunEvent {
  id: number
  event:
    | "run_started"
    | "node_started"
    | "node_completed"
    | "node_failed"
    | "waiting_for_user"
    | "run_completed"
    | "run_failed"
    | "run_cancelled"
  run_id: number
  node: string | null
  status: string
  timestamp: string
  duration_ms: number | null
  summary: Record<string, unknown>
  error_code: string | null
  error_message: string | null
  report_id: number | null
}

export interface MatchEvidenceRead extends EvidenceTrace {
  id: number
  detail_id: number
}

export interface MatchDetail {
  id: number
  report_id: number
  category: string
  requirement: string
  status: string
  score: number | null
  sort_order: number
  code: string | null
  severity: string | null
  explanation: string | null
  remediation: string | null
  matching_rule: string | null
  data: Record<string, unknown>
  evidence: MatchEvidenceRead[]
}

File: backend/app/schemas/matching.py
from __future__ import annotations

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, JsonValue, model_validator


class StrictSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")


class MatchStatus(StrEnum):
    PENDING = "PENDING"
    RUNNING = "RUNNING"
    SUCCESS = "SUCCESS"
    FAILED = "FAILED"
    CANCELLED = "CANCELLED"


class MatchPhase(StrEnum):
    VALIDATING_INPUT = "VALIDATING_INPUT"
    EXTRACTING_REQUIREMENTS = "EXTRACTING_REQUIREMENTS"
    NORMALIZING_SKILLS = "NORMALIZING_SKILLS"
    MATCHING_RULES = "MATCHING_RULES"
    MATCHING_SEMANTIC = "MATCHING_SEMANTIC"
    VERIFYING_EVIDENCE = "VERIFYING_EVIDENCE"
    CALCULATING_SCORE = "CALCULATING_SCORE"
    SAVING_REPORT = "SAVING_REPORT"
    COMPLETED = "COMPLETED"


class SkillMatchStatus(StrEnum):
    MATCHED = "MATCHED"
    PARTIAL = "PARTIAL"
    MISSING = "MISSING"
    UNKNOWN = "UNKNOWN"


class AvailabilityStatus(StrEnum):
    PROVIDED = "PROVIDED"
    UNKNOWN = "UNKNOWN"
    NOT_PROVIDED = "NOT_PROVIDED"


class RequirementInformationStatus(StrEnum):
    PROVIDED = "PROVIDED"
    NOT_REQUIRED = "NOT_REQUIRED"
    NOT_PROVIDED = "NOT_PROVIDED"
    UNPARSEABLE = "UNPARSEABLE"


class RiskSeverity(StrEnum):
    INFO = "INFO"
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    BLOCKING = "BLOCKING"


class RiskType(StrEnum):
    SKILL_GAP = "SKILL_GAP"
    BLOCKING_RISK = "BLOCKING_RISK"
    GENERAL_RISK = "GENERAL_RISK"


class VerificationStatus(StrEnum):
    CONFIRMED = "CONFIRMED"
    UNVERIFIED = "UNVERIFIED"
    CONFLICT = "CONFLICT"


class DimensionStatus(StrEnum):
    GOOD = "GOOD"
    PARTIAL = "PARTIAL"
    WEAK = "WEAK"
    UNKNOWN = "UNKNOWN"


class RecommendationLevel(StrEnum):
    STRONGLY_RECOMMENDED = "STRONGLY_RECOMMENDED"
    RECOMMENDED = "RECOMMENDED"
    CONSIDER = "CONSIDER"
    HIGH_RISK = "HIGH_RISK"
    NOT_RECOMMENDED = "NOT_RECOMMENDED"


class ReportConfidence(StrEnum):
    HIGH = "HIGH"
    MEDIUM = "MEDIUM"
    LOW = "LOW"
    VERY_LOW = "VERY_LOW"


class ConfidenceThresholds(StrictSchema):
    high_min: float = Field(default=0.8, ge=0, le=1)
    medium_min: float = Field(default=0.6, ge=0, le=1)
    low_min: float = Field(default=0.5, ge=0, le=1)

    @model_validator(mode="after")
    def validate_order(self) -> ConfidenceThresholds:
        if not self.high_min > self.medium_min > self.low_min:
            raise ValueError("confidence thresholds must descend from high to low")
        return self


class CompletenessThresholds(StrictSchema):
    unrestricted_min: float = Field(default=0.8, ge=0, le=1)
    recommended_cap_min: float = Field(default=0.6, ge=0, le=1)
    consider_cap_min: float = Field(default=0.5, ge=0, le=1)

    @model_validator(mode="after")
    def validate_order(self) -> CompletenessThresholds:
        if not self.unrestricted_min > self.recommended_cap_min > self.consider_cap_min:
            raise ValueError("completeness thresholds must descend from unrestricted")
        return self


class BlockingPolicyConfig(StrictSchema):
    policy_version: str = "blocking-policy-v1"
    work_eligibility_rule: str = "explicit_requirement_and_confirmed_conflict"
    mandatory_certificate_rule: str = "explicit_mandatory_and_confirmed_absence"
    mandatory_language_rule: str = "explicit_mandatory_and_confirmed_mismatch"
    experience_gap_rule: str = "explicit_minimum_and_gap_threshold"
    education_rule: str = "explicit_non_substitutable_minimum"
    explicit_requirement_markers: list[str] = Field(
        default_factory=lambda: [
            "must",
            "required",
            "mandatory",
            "must have",
            "minimum",
            "at least",
            "必须",
            "必需",
            "持有",
            "最低",
            "至少",
            "硬性要求",
            "不可替代",
        ]
    )
    preferred_requirement_markers: list[str] = Field(
        default_factory=lambda: [
            "preferred",
            "nice to have",
            "plus",
            "bonus",
            "优先",
            "加分",
            "更佳",
        ]
    )
    blocking_experience_gap_years: float = Field(default=3, gt=0, le=20)
    confidence_thresholds: ConfidenceThresholds = Field(default_factory=ConfidenceThresholds)
    completeness_thresholds: CompletenessThresholds = Field(default_factory=CompletenessThresholds)


class SemanticScoringConfig(StrictSchema):
    version: str = "semantic-v1"
    top_k: int = Field(default=5, ge=1, le=20)
    similarity_threshold: float = Field(default=0.25, ge=-1, le=1)
    similarity_ceiling: float = Field(default=1, gt=0, le=1)
    empty_score: float = Field(default=0, ge=0, le=100)


class HybridWeights(StrictSchema):
    deterministic: float = Field(default=0.7, ge=0, le=1)
    semantic: float = Field(default=0.3, ge=0, le=1)

    @model_validator(mode="after")
    def validate_total(self) -> HybridWeights:
        if abs(self.deterministic + self.semantic - 1) > 0.000001:
            raise ValueError("hybrid weights must total 1")
        return self


class ScoreWeights(StrictSchema):
    hard_skills: int = Field(default=35, ge=0, le=100)
    evidence_strength: int = Field(default=20, ge=0, le=100)
    experience_education: int = Field(default=15, ge=0, le=100)
    language_location_eligibility: int = Field(default=10, ge=0, le=100)
    user_preferences: int = Field(default=10, ge=0, le=100)
    other_conditions: int = Field(default=10, ge=0, le=100)

    @model_validator(mode="after")
    def validate_total(self) -> ScoreWeights:
        if sum(self.model_dump().values()) != 100:
            raise ValueError("matching dimension weights must total 100")
        return self


class EmbeddingConfigSnapshot(StrictSchema):
    provider: str
    model: str
    device: str
    normalize_embeddings: bool = True


class ScoringConfig(StrictSchema):
    version: str = "deterministic-v1"
    skill_dictionary_version: str = "skill-dictionary-v1"
    job_requirement_extractor_version: str = "job-requirements-v1"
    weights: ScoreWeights = Field(default_factory=ScoreWeights)
    partial_skill_credit: float = Field(default=0.5, ge=0, le=1)
    unknown_criterion_credit: float = Field(default=0.7, ge=0, le=1)
    required_skill_weight: int = Field(default=2, ge=1)
    preferred_skill_weight: int = Field(default=1, ge=1)
    blocking_policy: BlockingPolicyConfig | None = None
    base_rule_version: str | None = None
    semantic: SemanticScoringConfig | None = None
    hybrid_weights: HybridWeights | None = None
    embedding_config: EmbeddingConfigSnapshot | None = None


class SemanticEvidence(StrictSchema):
    rank: int = Field(ge=1)
    relation: str
    similarity: float = Field(ge=-1, le=1)
    score: float = Field(ge=0, le=100)
    resume_text: str
    job_text: str
    resume_section: str
    job_field: str
    page_number: int | None = Field(default=None, ge=1)
    paragraph_index: int | None = Field(default=None, ge=0)
    resume_metadata: dict[str, JsonValue]
    job_metadata: dict[str, JsonValue]


class SemanticSummary(StrictSchema):
    evidence_count: int = Field(ge=0)
    evaluated_resume_chunks: int = Field(ge=0)
    top_similarity: float | None = Field(default=None, ge=-1, le=1)
    relation_counts: dict[str, int] = Field(default_factory=dict)


class NormalizedSkill(StrictSchema):
    raw_name: str
    normalized_name: str
    category: str | None
    normalization_rule: str
    confidence: float = Field(ge=0, le=1)
    dictionary_version: str


class JobSkillRequirement(StrictSchema):
    raw_name: str
    normalized_name: str
    category: str | None
    required: bool
    weight: int = Field(ge=1)
    evidence_text: str
    source_field: str
    normalization_rule: str
    confidence: float = Field(ge=0, le=1)


class JobRequirements(StrictSchema):
    job_title: str
    industry: str | None = None
    salary_min: float | None = Field(default=None, ge=0)
    salary_max: float | None = Field(default=None, ge=0)
    skills: list[JobSkillRequirement]
    minimum_experience_years: float | None = Field(default=None, ge=0)
    experience_status: AvailabilityStatus
    graduate_exception: bool = False
    education_level: str | None = None
    education_status: AvailabilityStatus
    languages: list[str]
    language_status: AvailabilityStatus
    location: str | None = None
    location_status: AvailabilityStatus
    employment_type: str | None = None
    employment_status: AvailabilityStatus
    certificates: list[str]
    certificates_required: bool
    certificate_status: AvailabilityStatus
    work_eligibility: str | None = None
    work_eligibility_status: AvailabilityStatus
    experience_required_explicit: bool = False
    education_required_explicit: bool = False
    education_non_substitutable: bool = False
    language_required_explicit: bool = False
    language_preferred: bool = False
    certificate_required_explicit: bool = False
    certificate_preferred: bool = False
    warnings: list[str]
    extractor_version: str


class JobInformationAssessment(StrictSchema):
    statuses: dict[str, RequirementInformationStatus]
    job_information_completeness: float = Field(ge=0, le=1)
    evaluated_requirement_count: int = Field(ge=0)
    not_provided_requirement_count: int = Field(ge=0)
    unparseable_requirement_count: int = Field(ge=0)


class DimensionScore(StrictSchema):
    code: str
    label: str
    score: float = Field(ge=0, le=100)
    weight: int = Field(ge=0, le=100)
    weighted_score: float = Field(ge=0, le=100)
    explanation: str
    status: DimensionStatus = DimensionStatus.UNKNOWN
    evidence_keys: list[str] = Field(default_factory=list)
    gap_codes: list[str] = Field(default_factory=list)


class EvidenceTrace(StrictSchema):
    conclusion_key: str
    resume_version_id: int
    resume_skill_id: int | None
    resume_evidence: str | None
    resume_section: str | None
    source_type: str | None
    page_number: int | None = Field(default=None, ge=1)
    paragraph_index: int | None = Field(default=None, ge=0)
    source_location: str | None
    job_field: str
    job_snippet: str
    matching_rule: str
    confidence: float = Field(ge=0, le=1)


class SkillMatch(StrictSchema):
    normalized_name: str
    job_raw_name: str
    category: str | None
    requirement_type: str
    status: SkillMatchStatus
    score: float = Field(ge=0, le=1)
    matching_rule: str
    explanation: str
    resume_skill_id: int | None
    evidence_count: int = Field(ge=0)
    source_types: list[str]
    responsibility_keyword_overlap: bool


class RiskItem(StrictSchema):
    code: str
    title: str = ""
    severity: RiskSeverity
    explanation: str
    job_requirement: str | None
    resume_evidence: str | None
    remediation: str
    risk_type: RiskType = RiskType.GENERAL_RISK
    verification_status: VerificationStatus = VerificationStatus.UNVERIFIED

    @model_validator(mode="after")
    def populate_title(self) -> RiskItem:
        if not self.title:
            self.title = self.code.replace("_", " ").title()
        return self


class RecommendedAction(StrictSchema):
    code: str
    title: str
    explanation: str
    priority: RiskSeverity


class EvidenceCoverage(StrictSchema):
    total_requirements: int = Field(ge=0)
    verified: int = Field(ge=0)
    partially_supported: int = Field(ge=0)
    missing: int = Field(ge=0)
    unknown: int = Field(ge=0)
    coverage_percent: float = Field(ge=0, le=100)


class ResumeInputSnapshot(StrictSchema):
    resume_id: int
    version_id: int
    version_number: int
    is_confirmed: bool
    created_at: datetime


class JobInputSnapshot(StrictSchema):
    job_id: int
    title: str
    company: str
    location: str | None = None
    content_hash: str
    import_method: str | None = None
    description: str
    requirements: str | None = None
    responsibilities: str | None = None


class MatchInputSnapshot(StrictSchema):
    resume: ResumeInputSnapshot
    job: JobInputSnapshot


class MatchComputation(StrictSchema):
    rule_score: float = Field(ge=0, le=100)
    dimension_scores: list[DimensionScore]
    skill_matches: list[SkillMatch]
    risks: list[RiskItem]
    evidence_items: list[EvidenceTrace]
    recommendation: RecommendationLevel
    explanation: str
    recommended_actions: list[RecommendedAction]
    job_information_completeness: float | None = Field(default=None, ge=0, le=1)
    report_confidence: ReportConfidence | None = None
    report_confidence_score: float | None = Field(default=None, ge=0, le=1)
    completeness_warning: str | None = None
    evaluated_requirement_count: int = Field(default=0, ge=0)
    unknown_requirement_count: int = Field(default=0, ge=0)
    not_provided_requirement_count: int = Field(default=0, ge=0)
    recommendation_cap: RecommendationLevel | None = None
    policy_version: str | None = None
    policy_config_snapshot: BlockingPolicyConfig | None = None


class MatchCreate(StrictSchema):
    resume_version_id: int = Field(ge=1)
    job_id: int = Field(ge=1)
    scoring_version: str | None = Field(
        default=None,
        pattern=r"^(deterministic-v1(\.1)?|hybrid-v1)$",
    )


class MatchEvidenceRead(EvidenceTrace):
    id: int
    detail_id: int


class MatchDetailRead(StrictSchema):
    id: int
    report_id: int
    category: str
    requirement: str
    status: str
    score: float | None
    sort_order: int
    code: str | None
    severity: str | None
    explanation: str | None
    remediation: str | None
    matching_rule: str | None
    data: dict[str, JsonValue]
    evidence: list[MatchEvidenceRead]


class MatchReportRead(StrictSchema):
    id: int
    user_id: int
    resume_version_id: int
    job_id: int
    status: MatchStatus
    current_phase: MatchPhase
    rule_score: float | None
    semantic_score: float | None
    hybrid_score: float | None
    final_score: float | None
    scoring_version: str
    scoring_config_snapshot: ScoringConfig
    dimension_scores: list[DimensionScore]
    matched_skills: list[SkillMatch]
    partial_skills: list[SkillMatch]
    missing_skills: list[SkillMatch]
    hard_requirement_risks: list[RiskItem]
    evidence_items: list[EvidenceTrace]
    recommendation: RecommendationLevel | None
    explanation: str | None
    recommended_actions: list[RecommendedAction]
    job_requirements: JobRequirements | None
    phase_timings_ms: dict[str, int]
    duration_ms: int | None
    error_code: str | None
    error_message: str | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    job_information_completeness: float | None = Field(default=None, ge=0, le=1)
    report_confidence: ReportConfidence | None = None
    report_confidence_score: float | None = Field(default=None, ge=0, le=1)
    completeness_warning: str | None = None
    evaluated_requirement_count: int = Field(default=0, ge=0)
    unknown_requirement_count: int = Field(default=0, ge=0)
    not_provided_requirement_count: int = Field(default=0, ge=0)
    recommendation_cap: RecommendationLevel | None = None
    policy_version: str | None = None
    policy_config_snapshot: BlockingPolicyConfig | None = None
    embedding_model: str | None = None
    embedding_config_snapshot: EmbeddingConfigSnapshot | None = None
    semantic_config_snapshot: SemanticScoringConfig | None = None
    semantic_evidence: list[SemanticEvidence] = Field(default_factory=list)
    semantic_summary: SemanticSummary | None = None
    evidence_coverage: EvidenceCoverage
    input_snapshot: MatchInputSnapshot | None = None


class MatchCreateResult(StrictSchema):
    report: MatchReportRead
    reused: bool


class MatchListRead(StrictSchema):
    items: list[MatchReportRead]
    total: int
    offset: int
    limit: int


class MatchStatusRead(StrictSchema):
    id: int
    status: MatchStatus
    current_phase: MatchPhase
    phase_timings_ms: dict[str, int]
    duration_ms: int | None
    error_code: str | None
    error_message: str | None
